api/api.go

```diff
@@ -70,18 +72,39 @@ func New(
 	}, nil
 }
 
-// Register all APIs with the given router and return a mux.
-func (api *API) Register(r *route.Router, routePrefix string) *http.ServeMux {
+// Register all APIs. It registers APIv1 with the provided router directly. As
+// APIv2 works on the http.Handler level, this method also creates a new
+// http.ServeMux and then uses it to register both the provided router (to
+// handle "/") and APIv2 (to handle "<routePrefix>/api/v2"). The method returns
+// the newly created http.ServeMux.
+//
+// If the provided value for timeout is positive, it is enforced for all HTTP
```

Comment:
Instead of using a value smaller equal-zero as the null value, I would prefer using a pointer.  What do you think?

Reply:
I have just propagated the flag value (where, I think, we have no choice of giving non-positive values that meaning). With a pointer, we needed a separate translation from non-positive flag values to a nil pointer. And then let this function panic for non-positive values.

In different news: Isn't considering a non-positive timeout value as "no timeout" relatively common / intuitive?

Reply:
Given that we will probably have more optional configurations to come, how about adopting the `Option` pattern here, which we already use in `silence.go` and `nflog.go`. That enables us to do the validation at API creation time, giving users early feedback of invalid inputs at startup time:

```diff
diff --git a/api/api.go b/api/api.go
index 8241f730..8ee1fd01 100644
--- a/api/api.go
+++ b/api/api.go
@@ -48,6 +48,36 @@ func init() {
 type API struct {
 	v1 *apiv1.API
 	v2 *apiv2.API
+
+	// Timeout for HTTP GET requests.
+	getTimeout *time.Duration
+	// Maximum number of GET requests processed concurrently.
+	getConcurrency *int
+}
+
+// Option configures a new API implementation.
+type Option func(*API) error
+
+// WithGetTimeout sets the timeout for HTTP GET requests.
+func WithGetTimeout(t *time.Duration) Option {
+	return func(api *API) error {
+		api.getTimeout = t
+		return nil
+	}
+}
+
+// WithGetConcurrency sets the get concurrency for HTTP GET requests.
+func WithGetConcurrency(c *int) Option {
+	return func(api *API) error {
+		if c != nil && *c < 1 {
+			return fmt.Errorf(
+				"expected GetConcurrency API parameter to be greater or equal to one but got: '%v'",
+				c,
+			)
+		}
+		api.getConcurrency = c
+		return nil
+	}
 }
 
 // New creates a new API object combining all API versions.
@@ -57,6 +87,7 @@ func New(
 	sf func(model.Fingerprint) types.AlertStatus,
 	peer *cluster.Peer,
 	l log.Logger,
+	opts ...Option,
 ) (*API, error) {
 	v1 := apiv1.New(
 		alerts,
@@ -78,10 +109,18 @@ func New(
 		return nil, err
 	}
 
-	return &API{
+	api := &API{
 		v1: v1,
 		v2: v2,
-	}, nil
+	}
+
+	for _, o := range opts {
+		if err := o(api); err != nil {
+			return nil, err
+		}
+	}
+
+	return api, nil
 }
 
 // Register all APIs. It registers APIv1 with the provided router directly. As
@@ -100,9 +139,8 @@ func New(
 // concurrency limit.
 func (api *API) Register(
 	r *route.Router, routePrefix string,
-	timeout time.Duration, concurrency int,
 ) *http.ServeMux {
-	limiter := makeLimiter(timeout, concurrency)
+	limiter := makeLimiter(api.getTimeout, api.getConcurrency)
 
 	api.v1.Register(r.WithPrefix("/api/v1"))
 
@@ -144,14 +182,14 @@ func (api *API) Update(cfg *config.Config, resolveTimeout time.Duration) error {
 //
 // The returned middleware serves http.StatusServiceUnavailable (503) for requests that
 // would exceed the number.
-func makeLimiter(timeout time.Duration, concurrency int) func(http.Handler) http.Handler {
-	if concurrency < 1 {
-		concurrency = runtime.GOMAXPROCS(0)
-		if concurrency < 8 {
-			concurrency = 8
+func makeLimiter(timeout *time.Duration, concurrency *int) func(http.Handler) http.Handler {
+	if concurrency == nil {
+		*concurrency = runtime.GOMAXPROCS(0)
+		if *concurrency < 8 {
+			*concurrency = 8
 		}
 	}
-	inFlightSem := make(chan struct{}, concurrency)
+	inFlightSem := make(chan struct{}, *concurrency)
 
 	return func(h http.Handler) http.Handler {
 		concLimiter := http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
@@ -172,10 +210,10 @@ func makeLimiter(timeout time.Duration, concurrency int) func(http.Handler) http
 			}
 			h.ServeHTTP(rsp, req)
 		})
-		if timeout <= 0 {
+		if timeout == nil {
 			return concLimiter
 		}
-		return http.TimeoutHandler(concLimiter, timeout, fmt.Sprintf(
+		return http.TimeoutHandler(concLimiter, *timeout, fmt.Sprintf(
 			"Exceeded configured timeout of %v.\n", timeout,
 		))
 	}
diff --git a/cmd/alertmanager/main.go b/cmd/alertmanager/main.go
index 66a78934..dfffb4cf 100644
--- a/cmd/alertmanager/main.go
+++ b/cmd/alertmanager/main.go
@@ -150,8 +150,8 @@ func run() int {
 		externalURL    = kingpin.Flag("web.external-url", "The URL under which Alertmanager is externally reachable (for example, if Alertmanager is served via a reverse proxy). Used for generating relative and absolute links back to Alertmanager itself. If the URL has a path portion, it will be used to prefix all HTTP endpoints served by Alertmanager. If omitted, relevant URL components will be derived automatically.").String()
 		routePrefix    = kingpin.Flag("web.route-prefix", "Prefix for the internal routes of web endpoints. Defaults to path of --web.external-url.").String()
 		listenAddress  = kingpin.Flag("web.listen-address", "Address to listen on for the web interface and API.").Default(":9093").String()
-		getConcurrency = kingpin.Flag("web.get-concurrency", "Maximum number of GET requests processed concurrently. If negative or zero, the limit is GOMAXPROC or 8, whatever is larger.").Default("0").Int()
-		getTimeout     = kingpin.Flag("web.timeout", "Timeout for HTTP requests. If negative or zero, no timeout is set.").Default("0").Duration()
+		getConcurrency = kingpin.Flag("web.get-concurrency", "Maximum number of GET requests processed concurrently. If negative or zero, the limit is GOMAXPROC or 8, whatever is larger.").Int()
+		getTimeout     = kingpin.Flag("web.timeout", "Timeout for HTTP requests. If negative or zero, no timeout is set.").Duration()
 
 		clusterBindAddr = kingpin.Flag("cluster.listen-address", "Listen address for cluster.").
 				Default(defaultClusterAddr).String()
@@ -300,6 +300,7 @@ func run() int {
 		marker.Status,
 		peer,
 		log.With(logger, "component", "api"),
+		getTimeout, getConcurrency,
 	)
 
 	if err != nil {
@@ -400,7 +401,7 @@ func run() int {
 
 	ui.Register(router, webReload, logger)
 
-	mux := api.Register(router, *routePrefix, *getTimeout, *getConcurrency)
+	mux := api.Register(router, *routePrefix)
 
 	srv := http.Server{Addr: *listenAddress, Handler: mux}
 	srvc := make(chan struct{})
```

Reply:
Given that there seems to be no possibility in kingpin to tell, whether a user specified a flag or not, except comparing it to the default value :flushed:  , I guess this discussion is obsolete anyways and we can keep it as it is right now. I am very happy if you can proof this to be wrong @beorn7.

Reply:
Will look into Kingpin behavior.

I am not opposed to the options pattern, with the rationale you stated, that we will probably add more options later. However, my impression is that the "options struct" pattern has won over "options functions". And interestingly, AM uses the former for silences and the latter for nflog. Thus, neither of the two would be more consistent with the existing AM code. I would thus gravitate towards an options struct.

About the pointer vs. value: I'm not sold on the idea that the additional complication is worth it to use `nil` as a sentinel value for "default behavior". To prove my point, look at `*concurrency = runtime.GOMAXPROCS(0)`: This code line will result in a runtime error if `concurrency` is nil, but in your suggested code, you execute that code line precisely under the condition that `concurrency` is nil. Surely just an easy oversight, but it proves that the handling of a pointer is a bit trickier than a plain value. Plus, we need to come up with a meaning for a non-positive timeout and concurrency limits anyway. Alternatively, we could error out for non-positive timeout or concurrency limit, but that's additional validation. Finally, I'd like to have a way to specify the `--web.timeout` flag in a way that says "no timeout" other than _not_ specifying the flag at all. That's useful if you have a template in your configuration management that fills in a certain value for that flag. If you want to set no timeout, it's handy to specify "0" as a value rather than changing the whole template to not provide the flag at all.

In any case, I'll study Kingpin behavior, as said, and then come up with a suggestion.

Reply:
So yeah, kingpin flag values default to the zero value. Even the pointers just point to a variable with the zero value. Thus, I see no gain in introducing `nil` as a marker value for "default", even if I were of the opinion that that would be a good idea in general.

Reply:
I think it's best to give the `api.New` method an options struct as an argument. Working on it...